### PR TITLE
[Work in progress] Separation of concerns using React

### LINKED ISSUE #72: Implement conventions for panes that facilitate separation of concern
Some panes are difficult to understand today for various reasons: Some files are huge (close to 1000 SLOC), some functions are dependent on specific scoping, and some are very much dependent on intricate ways of manipulating the DOM.

We suggest introducing patterns/techniques and tools that are reused across the various panes to make it easier to get an overview of what a pane does.

A technology that might be of interest here is Web Components, that might allow us to create reusable components that can be reused across panes, perhaps even in other projects. A tool that might be of interest here is Stencil, which might make it easier for us to create web components (which could be easily ported to major frameworks as well).

*Action items*

- Find a pattern that faciliates a sensible separation of concern for panes
- Make the pattern a convention across all panes
- Optional: Make use of Web Components
- Optional: Make use of Stencil

### PR DESCRIPTION
**Pull request by Vinnl**
This is strongly based on #116. I created a Markdown pane with practically the same functionality and a similar architecture as in that PR, but this time using React. This primarily demonstrates that it is possible to use React in a single pane, without having to share a data model with the rest - it simply attaches to a DOM node, which is then provided to the OutlineManager.

The primary change is that there is no controller here. That said, there still is a "Container" component, which glues the data fetching to the view (ie hard to test, but logic that's not too complicated). That means that the view logic is relatively easy to test - an example is included.

I'll look at porting the TrustedApplications Pane next to get a feel for how this would work for a more involved Pane, but thought I'd submit this to give an initial view of what this might look like.

Fixes #72.

### CODE REVIEW

File: markdown/view.test.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/* eslint-env jest */
+import * as React from 'react'
+import {
+  render,
+  fireEvent
+} from '@testing-library/react'
+import { View } from './view'
+import { workaroundActError } from './actErrorWorkaround'
+
+workaroundActError()
```
Comment (megoth) [inline markdown/view.test.tsx:10]: Very nice that we can do snapshot testing!

File: markdown/index.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        newInstance
+      }))
+      .catch((err: any) => {
+        console.error('Error creating new instance of markdown file', err)
+        return options
+      })
+  },
+  render: (subject) => {
+    const container = document.createElement('div')
+    ReactDOM.render(<Container store={store} subject={subject}/>, container)
```
Comment (megoth) [inline markdown/index.tsx:29]: Very simple connection between panes and React... very interesting

File: markdown/container.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import * as React from 'react'
+import { loadMarkdown, saveMarkdown } from './service'
+import { View } from './view'
+import { ContainerProps } from '../types'
+
+export const Container: React.FC<ContainerProps> = (props) => {
```
Comment (megoth) [inline markdown/container.tsx:6]: Props should be typed of course, but looks like a reasonable way of passing data

File: markdown/view.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  const [rawText, setRawText] = React.useState(props.markdown)
+
+  function storeMarkdown () {
+    setPhase('saving')
+    props.onSave(rawText).then(() => {
+      setPhase('rendering')
+    })
+  }
+
+  if (phase === 'saving') {
```
Comment (megoth) [inline markdown/view.tsx:20]: I like this approach better than the conditionals in my experiment - I think this is easier to read.

**Review by megoth — COMMENTED** [review]
A lot of improvements here, and I'm more positive to React with this. Lets try a bit more experimenting and see where they take us.

File: markdown/container.tsx (hunk 3 above)
Comment (Vinnl) [reply markdown/container.tsx:6]: They are typed :) The generic passed to `React.FC` (i.e. `ContainerProps`) defines the types of the props. (I chose to extract them to a separate file here to encourage using the same API for all top-level containers - that would make it easier to potentially extract panes into separate apps, for example.)

**Comment by megoth** [comment]
This relates to https://github.com/solid/solid-panes/issues/72

**Comment by megoth** [comment]
Should keep this open until we've merged it into master.

**Comment by angelo-v** [comment]
@Vinnl I was going to work on a markdown pane and just found this! Any reason we should not merge it?

**Comment by Vinnl** [comment]
@angelo-v IIRC this was to prototype the use of React in solid-panes, which is quite a dependency. And the other reason would be the large number of conflicts :P

**Comment by angelo-v** [comment]
We included react already. My activtiystreams-pane is also react-based. So this should not be an issue. 
I guess the conflicts should also not be too hard, since it is quite well seperated

**Comment by Vinnl** [comment]
In that case I see no reason not to merge it :)

**Comment by angelo-v** [comment]
Ok, I will give it a try

**Comment by angelo-v** [comment]
I moved the code to a dedicated repo https://github.com/solid/markdown-pane

**Pull request closed without merging**